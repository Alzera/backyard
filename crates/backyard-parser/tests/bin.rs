use backyard_parser::parse_eval;

#[test]
fn basic() {
  let arena = bumpalo::Bump::new();
  let asts = parse_eval(&arena, "$a = 5 + 5;").unwrap();
  insta::assert_yaml_snapshot!(asts.serializable());
}

#[test]
fn chained() {
  let arena = bumpalo::Bump::new();
  let asts = parse_eval(
    &arena,
    "$this->a($a)
  ?? $this->b($b) ?? $this->c($c)
    ?? $this->d($d) ?? $this->e($e);"
  ).unwrap();
  insta::assert_yaml_snapshot!(asts.serializable());
}
