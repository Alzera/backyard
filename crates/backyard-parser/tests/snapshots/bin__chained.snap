---
source: crates/backyard-parser/tests/bin.rs
expression: asts
snapshot_kind: text
---
node_type: program
children:
  - node_type: bin
    left:
      node_type: call
      name:
        node_type: object_access
        object:
          node_type: this
          loc:
            start:
              line: 1
              column: 0
              offset: 0
            end:
              line: 1
              column: 0
              offset: 0
          leadings: ~
          trailings: ~
        property:
          node_type: identifier
          name: a
          loc:
            start:
              line: 1
              column: 7
              offset: 7
            end:
              line: 1
              column: 8
              offset: 8
          leadings: ~
          trailings: ~
        use_bracket: false
        is_nullsafe: false
        loc:
          start:
            line: 1
            column: 5
            offset: 5
          end:
            line: 1
            column: 7
            offset: 7
        leadings: ~
        trailings: ~
      arguments:
        - node_type: call_argument
          name: ~
          value:
            node_type: variable
            name:
              node_type: identifier
              name: a
              loc:
                start:
                  line: 1
                  column: 9
                  offset: 9
                end:
                  line: 1
                  column: 10
                  offset: 10
              leadings: ~
              trailings: ~
            loc:
              start:
                line: 1
                column: 9
                offset: 9
              end:
                line: 1
                column: 10
                offset: 10
            leadings: ~
            trailings: ~
          loc:
            start:
              line: 1
              column: 9
              offset: 9
            end:
              line: 1
              column: 9
              offset: 9
          leadings: ~
          trailings: ~
      loc:
        start:
          line: 1
          column: 8
          offset: 8
        end:
          line: 1
          column: 11
          offset: 11
      leadings: ~
      trailings: ~
    operator: "??"
    right:
      node_type: bin
      left:
        node_type: call
        name:
          node_type: object_access
          object:
            node_type: this
            loc:
              start:
                line: 2
                column: 5
                offset: 18
              end:
                line: 2
                column: 5
                offset: 18
            leadings: ~
            trailings: ~
          property:
            node_type: identifier
            name: b
            loc:
              start:
                line: 2
                column: 12
                offset: 25
              end:
                line: 2
                column: 13
                offset: 26
            leadings: ~
            trailings: ~
          use_bracket: false
          is_nullsafe: false
          loc:
            start:
              line: 2
              column: 10
              offset: 23
            end:
              line: 2
              column: 12
              offset: 25
          leadings: ~
          trailings: ~
        arguments:
          - node_type: call_argument
            name: ~
            value:
              node_type: variable
              name:
                node_type: identifier
                name: b
                loc:
                  start:
                    line: 2
                    column: 14
                    offset: 27
                  end:
                    line: 2
                    column: 15
                    offset: 28
                leadings: ~
                trailings: ~
              loc:
                start:
                  line: 2
                  column: 14
                  offset: 27
                end:
                  line: 2
                  column: 15
                  offset: 28
              leadings: ~
              trailings: ~
            loc:
              start:
                line: 2
                column: 14
                offset: 27
              end:
                line: 2
                column: 14
                offset: 27
            leadings: ~
            trailings: ~
        loc:
          start:
            line: 2
            column: 13
            offset: 26
          end:
            line: 2
            column: 16
            offset: 29
        leadings: ~
        trailings: ~
      operator: "??"
      right:
        node_type: bin
        left:
          node_type: call
          name:
            node_type: object_access
            object:
              node_type: this
              loc:
                start:
                  line: 2
                  column: 21
                  offset: 34
                end:
                  line: 2
                  column: 21
                  offset: 34
              leadings: ~
              trailings: ~
            property:
              node_type: identifier
              name: c
              loc:
                start:
                  line: 2
                  column: 28
                  offset: 41
                end:
                  line: 2
                  column: 29
                  offset: 42
              leadings: ~
              trailings: ~
            use_bracket: false
            is_nullsafe: false
            loc:
              start:
                line: 2
                column: 26
                offset: 39
              end:
                line: 2
                column: 28
                offset: 41
            leadings: ~
            trailings: ~
          arguments:
            - node_type: call_argument
              name: ~
              value:
                node_type: variable
                name:
                  node_type: identifier
                  name: c
                  loc:
                    start:
                      line: 2
                      column: 30
                      offset: 43
                    end:
                      line: 2
                      column: 31
                      offset: 44
                  leadings: ~
                  trailings: ~
                loc:
                  start:
                    line: 2
                    column: 30
                    offset: 43
                  end:
                    line: 2
                    column: 31
                    offset: 44
                leadings: ~
                trailings: ~
              loc:
                start:
                  line: 2
                  column: 30
                  offset: 43
                end:
                  line: 2
                  column: 30
                  offset: 43
              leadings: ~
              trailings: ~
          loc:
            start:
              line: 2
              column: 29
              offset: 42
            end:
              line: 2
              column: 32
              offset: 45
          leadings: ~
          trailings: ~
        operator: "??"
        right:
          node_type: bin
          left:
            node_type: call
            name:
              node_type: object_access
              object:
                node_type: this
                loc:
                  start:
                    line: 3
                    column: 7
                    offset: 54
                  end:
                    line: 3
                    column: 7
                    offset: 54
                leadings: ~
                trailings: ~
              property:
                node_type: identifier
                name: d
                loc:
                  start:
                    line: 3
                    column: 14
                    offset: 61
                  end:
                    line: 3
                    column: 15
                    offset: 62
                leadings: ~
                trailings: ~
              use_bracket: false
              is_nullsafe: false
              loc:
                start:
                  line: 3
                  column: 12
                  offset: 59
                end:
                  line: 3
                  column: 14
                  offset: 61
              leadings: ~
              trailings: ~
            arguments:
              - node_type: call_argument
                name: ~
                value:
                  node_type: variable
                  name:
                    node_type: identifier
                    name: d
                    loc:
                      start:
                        line: 3
                        column: 16
                        offset: 63
                      end:
                        line: 3
                        column: 17
                        offset: 64
                    leadings: ~
                    trailings: ~
                  loc:
                    start:
                      line: 3
                      column: 16
                      offset: 63
                    end:
                      line: 3
                      column: 17
                      offset: 64
                  leadings: ~
                  trailings: ~
                loc:
                  start:
                    line: 3
                    column: 16
                    offset: 63
                  end:
                    line: 3
                    column: 16
                    offset: 63
                leadings: ~
                trailings: ~
            loc:
              start:
                line: 3
                column: 15
                offset: 62
              end:
                line: 3
                column: 18
                offset: 65
            leadings: ~
            trailings: ~
          operator: "??"
          right:
            node_type: call
            name:
              node_type: object_access
              object:
                node_type: this
                loc:
                  start:
                    line: 3
                    column: 23
                    offset: 70
                  end:
                    line: 3
                    column: 23
                    offset: 70
                leadings: ~
                trailings: ~
              property:
                node_type: identifier
                name: e
                loc:
                  start:
                    line: 3
                    column: 30
                    offset: 77
                  end:
                    line: 3
                    column: 31
                    offset: 78
                leadings: ~
                trailings: ~
              use_bracket: false
              is_nullsafe: false
              loc:
                start:
                  line: 3
                  column: 28
                  offset: 75
                end:
                  line: 3
                  column: 30
                  offset: 77
              leadings: ~
              trailings: ~
            arguments:
              - node_type: call_argument
                name: ~
                value:
                  node_type: variable
                  name:
                    node_type: identifier
                    name: e
                    loc:
                      start:
                        line: 3
                        column: 32
                        offset: 79
                      end:
                        line: 3
                        column: 33
                        offset: 80
                    leadings: ~
                    trailings: ~
                  loc:
                    start:
                      line: 3
                      column: 32
                      offset: 79
                    end:
                      line: 3
                      column: 33
                      offset: 80
                  leadings: ~
                  trailings: ~
                loc:
                  start:
                    line: 3
                    column: 32
                    offset: 79
                  end:
                    line: 3
                    column: 32
                    offset: 79
                leadings: ~
                trailings: ~
            loc:
              start:
                line: 3
                column: 31
                offset: 78
              end:
                line: 3
                column: 34
                offset: 81
            leadings: ~
            trailings: ~
          loc:
            start:
              line: 3
              column: 20
              offset: 67
            end:
              line: 3
              column: 34
              offset: 81
          leadings: ~
          trailings: ~
        loc:
          start:
            line: 3
            column: 4
            offset: 51
          end:
            line: 3
            column: 34
            offset: 81
        leadings: ~
        trailings: ~
      loc:
        start:
          line: 2
          column: 18
          offset: 31
        end:
          line: 3
          column: 34
          offset: 81
      leadings: ~
      trailings: ~
    loc:
      start:
        line: 2
        column: 2
        offset: 15
      end:
        line: 3
        column: 34
        offset: 81
    leadings: ~
    trailings: ~
loc:
  start:
    line: 1
    column: 0
    offset: 0
  end:
    line: 3
    column: 35
    offset: 82
leadings: ~
trailings: ~
